src/k8s/pkg/snap/snap.go

```diff
@@ -354,4 +354,41 @@ func (s *snap) PreInitChecks(ctx context.Context, config types.ClusterConfig) er
 	return nil
 }
 
+func checkK8sServicePorts(config types.ClusterConfig, isControlPlane bool) error {
+	// NOTE(neoaggelos): in some environments the Kubernetes might hang when running for the first time
+	// This works around the issue by running them once during the install hook
+	ports := map[string]int{
+		// Default values from official Kubernetes documentation.
+		"kubelet":            10250,
+		"kubelet-healthz":    10248,
+		"kube-proxy-healhz":  10256,
+		"kube-proxy-metrics": 10249,
+		"k8s-dqlite":         config.Datastore.GetK8sDqlitePort(),
+		"loadbalancer":       config.LoadBalancer.GetBGPPeerPort(),
+	}
+
+	if isControlPlane {
+		ports["kube-apiserver"] = config.APIServer.GetSecurePort()
+		ports["kube-scheduler"] = 10259
+		ports["kube-controller-manager"] = 10257
+	} else {
+		ports["kube-apiserver-proxy"] = config.APIServer.GetSecurePort()
+	}
+
+	var allErrors []error
+	for service, port := range ports {
+		if port == 0 {
+			continue
+		}
+		if open, err := utils.IsLocalPortOpen(port); err != nil {
+			// Could not open port due to error.
+			allErrors = append(allErrors, fmt.Errorf("could not check port %d (needed by: %s): %w", port, service, err))
+		} else if open {
+			allErrors = append(allErrors, fmt.Errorf("port %d (needed by: %s) is already in use.", port, service))
```

Comment:
What does the final error message look like when multiple ports are in use?

Reply:
Something like this:

```
sudo k8s bootstrap
Bootstrapping the cluster. This may take a few seconds, please wait.
Error: Failed to bootstrap the cluster.

The error was: failed to POST /k8sd/cluster: failed to bootstrap new cluster: Failed to run post-bootstrap actions: pre-init checks failed for bootstrap node: Encountered error(s) while verifying port availability for Kubernetes services: Port 10250 (needed by: kubelet) is already in use.
Port 10249 (needed by: kube-proxy-metrics) is already in use.
```